buildpack/build_test.go

```diff
@@ -27,1335 +27,1323 @@ import (
+			logHandler     = memory.New()
+		)
+
+		it.Before(func() {
+			mockCtrl = gomock.NewController(t)
+			mockEnv = testmock.NewMockBuildEnv(mockCtrl)
+
+			var err error
+			tmpDir, err = ioutil.TempDir("", "lifecycle")
+			h.AssertNil(t, err)
+			stdout, stderr = &bytes.Buffer{}, &bytes.Buffer{}
+			platformDir = filepath.Join(tmpDir, "platform")
+			layersDir = filepath.Join(tmpDir, "launch")
+			appDir = filepath.Join(layersDir, "app")
+			h.Mkdir(t, layersDir, appDir, filepath.Join(platformDir, "env"))
+
+			storeDir, err = filepath.Abs(filepath.Join("testdata", strings.ToLower(kind), "by-id"))
+			h.AssertNil(t, err)
+
+			config = buildpack.BuildConfig{
+				AppDir:          appDir,
+				PlatformDir:     platformDir,
+				OutputParentDir: layersDir,
+				Out:             stdout,
+				Err:             stderr,
+				Logger:          &log.Logger{Handler: logHandler},
+			}
+
+			if kind == buildpack.KindBuildpack {
+				descriptor = buildpack.Descriptor{
+					API: api.Buildpack.Latest().String(),
+					Buildpack: buildpack.Info{
+						ID:       "A",
+						Version:  "v1",
+						Name:     "Buildpack A",
+						ClearEnv: false,
+						Homepage: "Buildpack A Homepage",
+					},
+					Dir: filepath.Join(storeDir, "A", "v1"),
+				}
+			} else {
+				descriptor = buildpack.Descriptor{
+					API: api.Buildpack.Latest().String(),
+					Extension: buildpack.Info{
+						ID:       "A",
+						Version:  "v1",
+						Name:     "Extension A",
+						ClearEnv: false,
+						Homepage: "Extension A Homepage",
+					},
+					Dir: filepath.Join(storeDir, "A", "v1"),
+				}
+			}
+		})
 
-func testBuild(t *testing.T, when spec.G, it spec.S) {
-	var (
-		bpTOML         buildpack.Descriptor
-		mockCtrl       *gomock.Controller
-		mockEnv        *testmock.MockBuildEnv
-		stdout, stderr *bytes.Buffer
-		tmpDir         string
-		platformDir    string
-		appDir         string
-		layersDir      string
-		buildpacksDir  string
-		config         buildpack.BuildConfig
-		logHandler     = memory.New()
-	)
-
-	it.Before(func() {
-		mockCtrl = gomock.NewController(t)
-		mockEnv = testmock.NewMockBuildEnv(mockCtrl)
-
-		var err error
-		tmpDir, err = ioutil.TempDir("", "lifecycle")
-		if err != nil {
-			t.Fatalf("Error: %s\n", err)
-		}
-		stdout, stderr = &bytes.Buffer{}, &bytes.Buffer{}
-		platformDir = filepath.Join(tmpDir, "platform")
-		layersDir = filepath.Join(tmpDir, "launch")
-		appDir = filepath.Join(layersDir, "app")
-		h.Mkdir(t, layersDir, appDir, filepath.Join(platformDir, "env"))
-
-		buildpacksDir, err = filepath.Abs(filepath.Join("testdata", "by-id"))
-		if err != nil {
-			t.Fatalf("Error: %s\n", err)
-		}
+		it.After(func() {
+			os.RemoveAll(tmpDir)
+			mockCtrl.Finish()
+		})
 
-		config = buildpack.BuildConfig{
-			AppDir:      appDir,
-			PlatformDir: platformDir,
-			LayersDir:   layersDir,
-			Out:         stdout,
-			Err:         stderr,
-			Logger:      &log.Logger{Handler: logHandler},
-		}
+		when("#Build", func() {
+			when("env", func() {
```

Comment:
This file was getting a little unwieldy, hence the huge diff. I'll add some comments to try to make it more understandable.